docs/man/man3/ekf.3
.TH "ekf" 3 "Sábado, 20 de Novembro de 2021" "Quadrirrotor" \" -*- nroff -*-
.ad l
.nh
.SH NAME
ekf \- Funções genéricas com álgebra de quatérnions e demais\&.  

.SH SYNOPSIS
.br
.PP
.SS "Estruturas de Dados"

.in +1c
.ti -1c
.RI "class \fBEKF\fP"
.br
.in -1c
.SS "Enumerações"

.in +1c
.ti -1c
.RI "enum \fB_status\fP { \fBBEGIN\fP, \fBREADY\fP, \fBFAIL\fP, \fBSTOPED\fP }"
.br
.in -1c
.SS "Funções"

.in +1c
.ti -1c
.RI "Eigen::MatrixXf \fBQ_r\fP (const Eigen::Ref< Eigen::Vector4f > &q)"
.br
.ti -1c
.RI "Eigen::MatrixXf \fBQ_l\fP (const Eigen::Ref< Eigen::Vector4f > &q)"
.br
.ti -1c
.RI "Eigen::Matrix3f \fBskew\fP (const Eigen::Vector3f &v)"
.br
.RI "Matrix antissimetrica\&. "
.ti -1c
.RI "Eigen::Matrix4f \fBS_r\fP (const Eigen::Ref< Eigen::Vector4f > q)"
.br
.RI "Calcula a Matriz \fIRight-Quaternion\fP a partir do quaternion q\&. "
.ti -1c
.RI "Eigen::Matrix4f \fBS_l\fP (const Eigen::Ref< Eigen::Vector4f > &q)"
.br
.RI "Calcula a Matriz \fILeft-Quaternion\fP a partir do quaternion q\&. "
.ti -1c
.RI "void \fBcomputeMcdFromQuaternion\fP (const Eigen::Vector4f &q, Eigen::Matrix3f &mcd)"
.br
.ti -1c
.RI "void \fBprintEigen\fP (const Eigen::MatrixXf &m)"
.br
.ti -1c
.RI "Eigen::Vector3f \fBcomputeVectorProjection\fP (const Eigen::Ref< Eigen::Vector3f > &v, const Eigen::Ref< Eigen::Vector3f > &u)"
.br
.ti -1c
.RI "Eigen::Vector4f \fBmultiplyQuaternions\fP (const Eigen::Ref< Eigen::VectorXf > &p, const Eigen::Ref< Eigen::VectorXf > &q)"
.br
.RI "Produto de quaternions aplicado\&. "
.ti -1c
.RI "void \fBquaternion2Euler\fP (Eigen::Ref< Eigen::Vector3f > _euler, const Eigen::Ref< const Eigen::Vector4f > &_q)"
.br
.ti -1c
.RI "void \fBcalc_phi\fP (float &x, const Eigen::Ref< const Eigen::Vector4f > &_q)"
.br
.RI "Calcula, a partir de um quatérnio de atitude, o valor do ângulo Phi em Radianos\&. "
.ti -1c
.RI "void \fBcalc_theta\fP (float &x, const Eigen::Ref< const Eigen::Vector4f > &_q)"
.br
.RI "Calcula, a partir de um quatérnio de atitude, o valor do ângulo Theta em Radianos\&. "
.ti -1c
.RI "void \fBcalc_psi\fP (float &x, const Eigen::Ref< const Eigen::Vector4f > &_q)"
.br
.RI "Calcula, a partir de um quatérnio de atitude, o valor do ângulo Psi em Radianos\&. "
.ti -1c
.RI "void \fBintegrationQuaternion\fP (Eigen::Vector4f &_quat, const Eigen::Ref< const Eigen::VectorXf > &_gyroscope, float &_dt)"
.br
.ti -1c
.RI "Eigen::Vector3f \fBrotateVectorWithQuaternion\fP (Eigen::Vector4f &q, Eigen::Vector3f &v_old)"
.br
.RI "Rotaciona um vetor tridimensional com um quaternion\&. "
.ti -1c
.RI "Eigen::Vector3f \fBrotateVectorWithQuaternion_Conjugate\fP (Eigen::Vector4f &q, Eigen::Vector3f &v_old)"
.br
.RI "Rotaciona um vetor tridimensional com um quaternion conjugado\&. "
.ti -1c
.RI "float \fBinvSqrt\fP (float _x)"
.br
.RI "Calcula o inverso do raiz de um float\&. "
.ti -1c
.RI "Eigen::Vector4f \fBq_theta\fP (float &phi)"
.br
.RI "Calcula o quaternion de atitude de rotação phi sobre o eixo x\&. "
.ti -1c
.RI "Eigen::Vector4f \fBq_phi\fP (float &theta)"
.br
.RI "Calcula o quaternion de atitude de rotação theta sobre o eixo y\&. "
.ti -1c
.RI "Eigen::Vector4f \fBq_psi\fP (float &psi)"
.br
.RI "Calcula o quaternion de atitude de rotação psi sobre o eixo z\&. "
.ti -1c
.RI "void \fBcomputeQuaternionFromMCDMarkley\fP (Eigen::Matrix3f &m, Eigen::Ref< Eigen::Vector4f > q)"
.br
.ti -1c
.RI "Eigen::VectorXf \fBRK\fP (float h, const Eigen ::Ref< Eigen::VectorXf > &xk, const Eigen ::Ref< Eigen::VectorXf > &u, Eigen::VectorXf(fn)(Eigen::VectorXf x, Eigen::VectorXf u))"
.br
.ti -1c
.RI "void \fBconjugate\fP (Eigen::Ref< Eigen::Vector4f > _q)"
.br
.RI "Conjuga o quaternion _q\&. "
.ti -1c
.RI "Eigen::Matrix4f \fBS_r\fP (const Eigen::Ref< Eigen::Vector4f > &q)"
.br
.ti -1c
.RI "void \fBcomputeQuaternionFromMCDShepperd\fP (Eigen::Matrix3f &m, Eigen::Vector4f &q)"
.br
.ti -1c
.RI "Eigen::VectorXf \fBRK\fP (float h, const Eigen ::Ref< Eigen::VectorXf > &xk, const Eigen ::Ref< Eigen::VectorXf > &u, Eigen::VectorXf fn(Eigen::VectorXf x, Eigen::VectorXf u))"
.br
.in -1c
.SS "Variáveis"

.in +1c
.ti -1c
.RI "\fBAQUA\fP \fBaqua\fP"
.br
.ti -1c
.RI "const float \fBTOLERANCE\fP = 1e\-6"
.br
.in -1c
.SH "Descrição detalhada"
.PP 
Funções genéricas com álgebra de quatérnions e demais\&. 
.SH "Enumerações"
.PP 
.SS "enum \fBekf::_status\fP"
Enum Status Filter 
.PP
\fBEnumeradores\fP
.in +1c
.TP
\fB\fIBEGIN \fP\fP
.TP
\fB\fIREADY \fP\fP
.TP
\fB\fIFAIL \fP\fP
.TP
\fB\fISTOPED \fP\fP
.PP
Definição na linha 44 do arquivo EKF\&.h\&.
.PP
.nf
44 { BEGIN, READY, FAIL, STOPED };
.fi
.SH "Funções"
.PP 
.SS "void ekf::calc_phi (float & x, const Eigen::Ref< const Eigen::Vector4f > & _q)"

.PP
Calcula, a partir de um quatérnio de atitude, o valor do ângulo Phi em Radianos\&. 
.PP
\fBParâmetros\fP
.RS 4
\fIx\fP Referencia a ser sobrescrita\&. -pi < x < pi 
.br
\fI_q\fP Quaternion de atitude 
.RE
.PP

.PP
Definição na linha 173 do arquivo GRUPO_QUAT\&.cpp\&.
.PP
.nf
173                                                                  {
174   x = -atan2f(_q(2) * _q(3) - _q(0) * _q(1),
175               _q(0) * _q(0) + _q(3) * _q(3) - 0\&.5f);
176 }
.fi
.PP
Referenciado(a) por quaternion2Euler()\&.
.SS "void ekf::calc_psi (float & x, const Eigen::Ref< const Eigen::Vector4f > & _q)"

.PP
Calcula, a partir de um quatérnio de atitude, o valor do ângulo Psi em Radianos\&. 
.PP
\fBParâmetros\fP
.RS 4
\fIx\fP Referencia a ser sobrescrita\&. -pi < x < pi 
.br
\fI_q\fP Quaternion de atitude 
.RE
.PP

.PP
Definição na linha 194 do arquivo GRUPO_QUAT\&.cpp\&.
.PP
.nf
194                                                                  {
195   x = -atan2f(_q(1) * _q(2) - _q(0) * _q(3),
196               _q(0) * _q(0) + _q(1) * _q(1) - 0\&.5f);
197 }
.fi
.PP
Referenciado(a) por Controle::Controle() e quaternion2Euler()\&.
.SS "void ekf::calc_theta (float & x, const Eigen::Ref< const Eigen::Vector4f > & _q)"

.PP
Calcula, a partir de um quatérnio de atitude, o valor do ângulo Theta em Radianos\&. 
.PP
\fBParâmetros\fP
.RS 4
\fIx\fP Referencia a ser sobrescrita\&. -pi/2 < x < pi/2 
.br
\fI_q\fP Quaternion de atitude 
.RE
.PP

.PP
Definição na linha 184 do arquivo GRUPO_QUAT\&.cpp\&.
.PP
.nf
184                                                                    {
185   x = -asinf(-2\&.0f * (_q(0) * _q(2) + _q(1) * _q(3)));
186 }
.fi
.PP
Referenciado(a) por quaternion2Euler()\&.
.SS "void ekf::computeMcdFromQuaternion (const Eigen::Vector4f & q, Eigen::Matrix3f & mcd)"
Calcula a matriz de cossenos ditores a parir do quaternion de atitude 
.PP
\fBParâmetros\fP
.RS 4
\fIq\fP quaternion de atitude 
.br
\fImcd\fP Matriz de cossenos diretores a ser atualizada (Referencia) 
.RE
.PP

.PP
Definição na linha 102 do arquivo GRUPO_QUAT\&.cpp\&.
.PP
.nf
102                                                                           {
103   mcd = Eigen::Matrix3f::Identity() * (2\&.0f * q(0) * q(0) - 1\&.0f) +
104         2\&.0f * (q\&.tail(3) * q\&.tail(3)\&.transpose()) +
105         2\&.0f * q(0) * skew(q\&.tail(3));
106 }
.fi
.PP
Referencias skew()\&.
.SS "void ekf::computeQuaternionFromMCDMarkley (Eigen::Matrix3f & m, Eigen::Ref< Eigen::Vector4f > q)"
Extrai o quatérnio a partir da matrix de cossenos diretores\&. 
.PP
\fBParâmetros\fP
.RS 4
\fIm\fP Matrix de cossenos diretores\&. 
.br
\fIq\fP Quaternion de atitude\&. 
.RE
.PP
\fBObservação\fP
.RS 4
Utiza o algoritmo BUG 
.RE
.PP

.PP
Definição na linha 350 do arquivo GRUPO_QUAT\&.cpp\&.
.PP
.nf
351                                                                 {
352   Eigen::Vector4f indexVector;
353   float _trace = m\&.trace();
354   indexVector << _trace, m\&.diagonal();
355   Eigen::Vector4f::Index max_index;
356   indexVector\&.maxCoeff(&max_index);
357   switch (max_index) {
358   case 0:
359     q(0) = 1\&.0f + _trace;
360     q(1) = m(2, 1) - m(1, 2);
361     q(2) = m(0, 2) - m(2, 0);
362     q(3) = m(1, 0) - m(0, 1);
363     break;
364   case 1:
365     q(0) = m(2, 1) - m(1, 2);
366     q(1) = 1\&.0f + m(0, 0) - m(1, 1) - m(2, 2);
367     q(2) = m(0, 1) + m(1, 0);
368     q(3) = m(2, 0) + m(0, 2);
369     break;
370   case 2:
371     q(0) = m(2, 0) - m(0, 2);
372     q(1) = m(0, 1) + m(1, 0);
373     q(2) = 1\&.0f - m(0, 0) + m(1, 1) - m(2, 2);
374     q(3) = m(1, 2) + m(2, 1);
375     break;
376   default:
377     q(0) = m(1, 0) - m(0, 1);
378     q(1) = m(2, 0) + m(0, 2);
379     q(2) = m(2, 1) + m(1, 2);
380     q(3) = 1\&.0f - m(0, 0) - m(1, 1) + m(2, 2);
381     break;
382   }
383   q\&.normalize();
384 }
.fi
.PP
Referencias drone::m\&.
.SS "void ekf::computeQuaternionFromMCDShepperd (Eigen::Matrix3f & m, Eigen::Vector4f & q)"

.SS "Eigen::Vector3f ekf::computeVectorProjection (const Eigen::Ref< Eigen::Vector3f > & v, const Eigen::Ref< Eigen::Vector3f > & u)"
Rejeita a projecao do vetor v sobre u \&. 
.PP
\fBParâmetros\fP
.RS 4
\fIu\fP Vetor tridimensional 
.br
\fIv\fP Vetor tridimensional 
.RE
.PP
\fBRetorna\fP
.RS 4
v sem a projeção de v sobre u\&. * \fCVector Rejection\fP [Vector Rejection ] 
.RE
.PP

.PP
Definição na linha 130 do arquivo GRUPO_QUAT\&.cpp\&.
.PP
.nf
131                                                                           {
132 
133   return (v - v\&.dot(u) * u);
134 }
.fi
.SS "void ekf::conjugate (Eigen::Ref< Eigen::Vector4f > _q)"

.PP
Conjuga o quaternion _q\&. 
.PP
\fBParâmetros\fP
.RS 4
\fI_q\fP 
.RE
.PP

.PP
Definição na linha 400 do arquivo GRUPO_QUAT\&.cpp\&.
.PP
.nf
400 { _q\&.tail(3) *= -1\&.0f; }
.fi
.SS "void ekf::integrationQuaternion (Eigen::Vector4f & _quat, const Eigen::Ref< const Eigen::VectorXf > & _gyroscope, float & _dt)"
Integracao de quaternion com mapa exponecial\&. 
.PP
\fBParâmetros\fP
.RS 4
\fI_quat\fP Quaternion q_k 
.br
\fI_giroscopio\fP medidas do giroscópio [Radianos] 
.br
\fI_dt\fP Passo de integração\&.
.RE
.PP
\fBObservação\fP
.RS 4
computa norma para evitar instabilidade
.RE
.PP

.PP
Definição na linha 198 do arquivo GRUPO_QUAT\&.cpp\&.
.PP
.nf
200                                        {
208   float _gyroscopeNorm = _gyroscope\&.norm();
209 
210   if (_gyroscopeNorm > TOLERANCE) {
211     Eigen::Vector4f expMap;
212     float _dtGyroscopeNormHalf = 0\&.5f * _dt * _gyroscopeNorm;
213     expMap(0) = cosf(_dtGyroscopeNormHalf);
214     expMap\&.tail(3) = _gyroscope / _gyroscopeNorm * sinf(_dtGyroscopeNormHalf);
215     _quat = multiplyQuaternions(_quat, expMap);
216   }
217 }
.fi
.PP
Referencias multiplyQuaternions() e TOLERANCE\&.
.SS "float ekf::invSqrt (float _x)"

.PP
Calcula o inverso do raiz de um float\&. 
.PP
\fBObservação\fP
.RS 4
Para mais detalhes consulte \fCFast inverse square-root\fP 
.RE
.PP
\fBParâmetros\fP
.RS 4
\fI_x\fP 
.RE
.PP
\fBRetorna\fP
.RS 4
1/sqrt(_x) 
.RE
.PP

.PP
Definição na linha 256 do arquivo GRUPO_QUAT\&.cpp\&.
.PP
.nf
256                         {
257   float halfx = 0\&.5f * _x;
258   union {
259     float _y;
260     uint32_t _i;
261   } conv = {\&._y = _x};
262   // float _y = _x;
263   // long i = *(long *)&_y;
264   conv\&._i = 0x5f3759df - (conv\&._i >> 1);
265   // _y = *(float *)&i;
266   conv\&._y *= (1\&.5f - (halfx * conv\&._y * conv\&._y));
267   // segunda iteracao\&.
268   // _y = _y * (1\&.5f - (halfx * _y * _y));
269   return conv\&._y;
270 }
.fi
.PP
Referenciado(a) por AQUA::computeQuaternionMag()\&.
.SS "Eigen::Vector4f ekf::multiplyQuaternions (const Eigen::Ref< Eigen::VectorXf > & p, const Eigen::Ref< Eigen::VectorXf > & q)"

.PP
Produto de quaternions aplicado\&. 
.PP
\fBParâmetros\fP
.RS 4
\fIp\fP Quaternion 
.br
\fIq\fP Vetor tridimensional 
.RE
.PP
\fBRetorna\fP
.RS 4
Quaternion product pq 
.RE
.PP

.PP
Definição na linha 142 do arquivo GRUPO_QUAT\&.cpp\&.
.PP
.nf
143                                                                       {
144   Eigen::Vector4f pq;
145   pq(0) = p(0) * q(0) - p(1) * q(1) - p(2) * q(2) - p(3) * q(3);
146   pq(1) = p(0) * q(1) + p(1) * q(0) + p(2) * q(3) - p(3) * q(2);
147   pq(2) = p(0) * q(2) - p(1) * q(3) + p(2) * q(0) + p(3) * q(1);
148   pq(3) = p(0) * q(3) + p(1) * q(2) - p(2) * q(1) + p(3) * q(0);
149   return pq;
150 }
.fi
.PP
Referenciado(a) por AQUA::computeAQUAQuaternion() e integrationQuaternion()\&.
.SS "void ekf::printEigen (const Eigen::MatrixXf & m)"
Imprime um dado Eigen\&. Matriz ou vetor\&. 
.PP
\fBParâmetros\fP
.RS 4
\fIm\fP dado a ser impresso\&. 
.RE
.PP

.PP
Definição na linha 111 do arquivo GRUPO_QUAT\&.cpp\&.
.PP
.nf
111                                         {
112   int c = m\&.cols();
113   int l = m\&.rows();
114   for (int i = 0; i < l; i++) {
115     for (int j = 0; j < c; j++) {
116       // printf("%f\t", m(i, j));
117       printf(",%f", m(i, j));
118     }
119     printf("%s\n", ",");
120   }
121 }
.fi
.PP
Referencias drone::m\&.
.SS "Eigen::MatrixXf ekf::Q_l (const Eigen::Ref< Eigen::Vector4f > & q)"
Matrix Left-Quaternion Q 
.PP
\fBParâmetros\fP
.RS 4
\fIq\fP quaternion de atitude 
.RE
.PP
\fBRetorna\fP
.RS 4
Matrix Left Quarternio Q 
.RE
.PP

.PP
Definição na linha 45 do arquivo GRUPO_QUAT\&.cpp\&.
.PP
.nf
45                                                     {
46 
47   Eigen::MatrixXf M = Eigen::MatrixXf::Zero(4, 3);
48   M\&.row(0) = -q\&.tail(3);
49   M\&.bottomLeftCorner(3, 3) =
50       q(0) * Eigen::Matrix3f::Identity() + skew(q\&.tail(3));
51   return M;
52 }
.fi
.PP
Referencias skew()\&.
.PP
Referenciado(a) por Controle::computeRotationalTarget()\&.
.SS "Eigen::Vector4f ekf::q_phi (float & theta)"

.PP
Calcula o quaternion de atitude de rotação theta sobre o eixo y\&. 
.PP
\fBParâmetros\fP
.RS 4
\fItheta\fP 
.RE
.PP
\fBRetorna\fP
.RS 4
Eigen::Vector4f Quaternion de rotacao theta 
.RE
.PP

.PP
Definição na linha 286 do arquivo GRUPO_QUAT\&.cpp\&.
.PP
.nf
286                                   {
287   return Eigen::Vector4f{cosf(0\&.5f * theta), 0, sinf(0\&.5f * theta), 0};
288 }
.fi
.SS "Eigen::Vector4f ekf::q_psi (float & psi)"

.PP

lib/EKF/EKF.cpp
/**
 * @author: Roney Silva (roney)
 * @date:   16-Aug-2021
 * Email:  roneyddasilva@gmail.com
 * Project: quadrirrotorUFABC
 * @file: EKF.cpp
 * Last modified by:   roney
 * Last modified time: 01-Sep-2021
 * @brief Classe que implementa o filtro de Kalman estendido
 */

#include "EKF.h"
#ifdef __XTENSA__
#include "GPS.h"

#include "IMU.h"
#include "TIME.h"
#include <Arduino.h>
#endif
// #include "../AQUA/AQUA.h"
// #include "../GRUPO_QUAT/GRUPO_QUAT.h"

namespace ekf {
/**
 * Classe do filtro de Kalman.
 */
#ifdef __XTENSA__
IMU imu(Wire, 0x68);
TIME timeClass;
TIME timeClassLoop;
#endif
AQUA aqua;
EKF::EKF() {}
/**
 * @brief Inicializacao do filtro de Kalman
 * @details Inicializa e configura:o GPS, a classe timeClass, instancia os
 * sensores #accel, #gyro e #mag; e Configura os valores padrões da matrizes #F,
 * #G e #H.
 */
void EKF::begin(float _dt) {
#ifdef __XTENSA__
  printf("Nucleo do Filtro: %d\n", xPortGetCoreID());
#endif
  status = BEGIN;
  this->dt = _dt;
#ifdef __XTENSA__
  gpsSetup();

  timeClass.begin();
  // timeClassLoop.begin();

#endif

  // Instancia o vetor de medidas na biblioteca da IMU.
  /*!HACK: @hack #accel, #gyro e #mag são ponteiros. Quando as medidas são
   * atualizadas na biblioteca IMU, o vetor é modificado na EKF. */
#ifdef __XTENSA__
  imu.begin(accel, gyro, mag);
  // referencia o bias da classe IMU.
  biasGyro = &imu._biasGyro;
  biasAccel = &imu._biasAccel;
#endif
  // Atualiza valor do bias do giroscópio com uma média de amostras
  // valores constantes das matriz F
  mF.block<3, 3>(0, 3) = -0.5f * Eigen::Matrix3f::Identity();
  mF.block<3, 3>(12, 6) = Eigen::Matrix3f::Identity();
  // Valores constantes da matriz G
  G.topLeftCorner(3, 3) = -0.5f * Eigen::Matrix3f::Identity();
  G.block<3, 3>(3, 3) = Eigen::Matrix3f::Identity();
  G.block(6, 6, 3, 3) = -Eigen::Matrix3f::Identity();
  G.block<3, 3>(9, 9) = Eigen::Matrix3f::Identity();
  // matrix de covariâncias do processo continuo
  Eigen::MatrixXf Q = Eigen::MatrixXf::Identity(12, 12);
  Eigen::Matrix3f _I = Eigen::Matrix3f::Identity();
  Q.block<3, 3>(0, 0) = _I * _dt * _dt * sigma_giro_squared;
  Q.block<3, 3>(3, 3) = _I * _dt * sigma_bias_giro_squared;
  Q.block<3, 3>(6, 6) = _I * _dt * _dt * sigma_accel_squared;
  Q.block<3, 3>(9, 9) = _I * _dt * sigma_bias_accel_squared;
  Q_k = G * Q * G.transpose();
  R.block<4, 4>(0, 0) = Eigen::Matrix4f::Identity() * sigma_aqua_squared;

#ifndef __XTENSA__
  // Q_k = csvLeia<Eigen::MatrixXf>("Q");
  // R = csvLeia<Eigen::MatrixXf>("R");
#endif
  // condições iniciais da posicao
  aqua.begin(qObs, accel, mag);
#ifdef __XTENSA__
  updateBaseRef();
  imu.readSensor();
#endif
  // valor inicial a partir da medida
  aqua.computeAQUAQuaternion();
  q = qObs;
  // aqua.qAlignment = qObs;
  status = READY;
}
/*!
 *  Loop do Filtro de Kalman
 */
void EKF::loopEKF() {
  // le os sensores
  // timeClassLoop.computeElapsedTime();
  updateOfMeasurements();
  predictionStage();
  updateStage();
  updateStates();
  // dtLoop = timeClassLoop.computeElapsedTime();
}

/**
 *  Integra os estados e a matriz de covariancia a priori do filtro.
 */
void EKF::predictionStage() {
  // Integracao dos estados
  integrationOfStates();
  updateF();
  updateFi();
  // atualiza a matriz de covariâncias
  P = fi * P * fi.transpose() + Q_k;
}
/**
 * Atualiza a medicao a partir dos sensores da unidade Margin e GPS.
 * @details Faz a leitura dos sensores da MARG atualizando os vetores #accel,
 * #gyro e #mag e ElapsedTime #dt.
 * Atualiza as medidas do GPS quando estiverem disponíveis, o valor da variável
 * #p e bloco da matriz H correspondente.
 */
void EKF::updateOfMeasurements() {
#ifdef __XTENSA__
  // Faz a leitura dos sensores da MARG atualizando os vetores #accel, #gyro e
  imu.readSensor();
  // tempo decorrido entre medidas a medida anterior e a atual.
  // dt = timeClass.computeElapsedTime();
  // Etapa de predição. Verifica se há medidas novas a partir do GPS. A função
  // processoGPS() retorna a MT_NONE, MT_NAV_POSLLH, MT_NAV_VELNED ou MT_NAV_PVT
  // dependendo da configuração no cabeçalho GPS.h.
  messageGPS = processGPS();
  /*! @note MT_NONE é avalidado primeiro uma vez que a frequência do GPS é
menor que a da MARG.*/
  /*! @todo É possível que essas contas sejam realizadas no
   * cabeçalho, eliminando-se alguns condicionais abaixo.*/
  // p e n corresponde aos numeros de medidas e estados.
  if (messageGPS == MT_NONE) {
    // não ha observações de posição ou velocidade
    p = pAtitude;
  } else if (messageGPS == MT_NAV_POSECEF) {
    // há observação de posicao
    p = pFull;
    // Atualiza o erro de medida na velocidade linear
    updateNedPos();
    // Atualiza matriz covariancia da matriz de observação da posição
    R.block<3, 3>(4, 4) = sigma_pos_squared * Eigen::Matrix3f::Identity();
    deltaYObs.tail(3) = rNEDgps - rNED;
  } else if (messageGPS == MT_NAV_PVT) {
    p = pFull;
    // há observações de velocidade
    updateNedVel();
    // Atualiza matriz covariancia da matriz de observação da velocidade
    R.block<3, 3>(4, 4) = sigma_vel_squared * Eigen::Matrix3f::Identity();
    deltaYObs.tail(3) = drNEDgps - drNED;
  }
#endif
  // computa o quaternion de observação com o algoritmo #AQUA.
  aqua.computeAQUAQuaternion();
}
/**
 * @brief Atualiza matrix lineariza de estados #F e a matrix #G.
 */
void EKF::updateF() {
  /*! @hack Apenas parte da matrix é atualizada*/
  mF.topLeftCorner(3, 3) = -skew(gyro);
  // TEST: Matrix mF atualizada em todos os loops
  mF.block<3, 3>(6, 0) = -2.0 * mcd * skew(accel);
  mF.block<3, 3>(6, 9) = -mcd;
}
/**
 * @brief
 * @todo Oportunidade de otimização. Talves, na programação, a matrix F
 * não precise existir e seja absorvida pela matrix fi.
 */
void EKF::updateFi() {
  // fi = I+dt * F
  fi = Eigen::MatrixXf::Identity(nFull, nFull) + (mF * (0.5f * dt) + mF) * dt;
}

/**
 * Integra os estado #q e #r;.
 */
void EKF::integrationOfStates() {
  // Quatérnio com Euler de primeira ordem
  integrationQuaternion(q, gyro, dt);
  // atualiza matriz de rotação com o quaternio predito
  computeMcdFromQuaternion(q, mcd);
  Eigen::Vector3f _accel_g = dt * (mcd * accel - navegationGravity);
  // integracao da Posicao NED
  rNED += (drNED + 0.5f * _accel_g) * dt;
  // integracao da Velocidade
  drNED += _accel_g;
}
/**
 * @brief Etapa de atualização.
 */
void EKF::updateStage() {
  // Ganho de Kalman
  deltaYObs.head(4) = qObs - q;
  updateHq();
  K.leftCols(p) =
      P * H.topRows(p).transpose() *
      (H.topRows(p) * P * H.topRows(p).transpose() + R.topLeftCorner(p, p))
          .inverse();
  // Atualiza a matriz de covariância.
  P -= K.leftCols(p) * H.topRows(p) * P;
  // Computa o erro de estimação #deltaChi
  deltaChi.segment(1, n) = K.leftCols(p) * deltaYObs.head(p);
}
/**
 * Atualiza os estados predidos com os erros estimados.
 * @note Quando há medidas do GPS o bias do acelerômetro, a velocidade NED e a
 * posição Geodésica são atualizadas.
 */
void EKF::updateStates() {
  // calcula o escalar do erro de estimação.
  deltaChi(0) = sqrtf(1.0f - deltaChi.segment<3>(1).squaredNorm());
  // Atualiza a orientação.
  q = multiplyQuaternions(q, deltaChi.head(4));
  // Atualiza o bias do giroscópio.
  *biasGyro += deltaChi.segment<3>(4);
  /*! @attention dcm é uma referência. */
  // Corrige a velocidade de translacao.
  drNED += deltaChi.segment<3>(7);
  // Corrige o bias do acelerometro.
  *biasAccel += deltaChi.segment<3>(10);
  // Corrige a posição Geodésica
  rNED += deltaChi.segment<3>(13);
}
/**
 * @brief Atuliza o bloco pertinente á atitude na matriz #H
 */
void EKF::updateHq() { H.block<4, 3>(0, 0) = Q_l(q); }
/**
 * @brief Atualiza parte relativa á posição da matriz de sensitividade.
 */
void EKF::updateHp() {
  H.block<3, 3>(4, 6) = Eigen::Matrix3f::Zero();
  H.block<3, 3>(4, 12) = Eigen::Matrix3f::Identity();
}
/**
 * @brief Atualiza parte relatica á velocidade da matriz de sensitividade.
 */
void EKF::updateHv() {
  H.block<3, 3>(4, 6) = Eigen::Matrix3f::Identity();
  H.block<3, 3>(4, 12) = Eigen::Matrix3f::Zero();
}

#ifdef __XTENSA__
/**
 * @brief Função que armazena as coordenas da base.
 * @details É atualizada a estrutura #base.
 */
void EKF::updateBaseRef() {
  blindLED(2);
  // raio Meridional
  base.RotEcef2ned = Eigen::Matrix3f::Identity();
  base.rECEF[0] = 0;
  base.rECEF[1] = 0;
  base.rECEF[2] = 0;
  int _mensagem = processGPS();
  ubxMessage.navPosecef.pAcc = 1000;
  while (ubxMessage.navPosecef.pAcc > 300) {
    while (true) {
      _mensagem = processGPS();
      if (_mensagem == MT_NAV_POSECEF)
        break;
    }
#ifdef DEBUG_EKF
    printf("mensagem: %d\t%lu\n", _mensagem, ubxMessage.navPosecef.pAcc);
#endif
    vTaskDelay(100);
    // mensagem em cm/s e em long para subtrair na transformacao ECEF->NED
    base.rECEF[0] = ubxMessage.navPosecef.ecefX;
    base.rECEF[1] = ubxMessage.navPosecef.ecefY;
    base.rECEF[2] = ubxMessage.navPosecef.ecefZ;
  }
  while (ubxMessage.navPvt.fixType != 3) {
    while (true) {
      _mensagem = processGPS();
      if (_mensagem == 2)
        break;
    }
  }
#ifdef DEBUG_EKF
  printf("%d\n", ubxMessage.navPvt.fixType);
#endif
  base.lat = (float)ubxMessage.navPvt.lat * DEG_TO_RAD_BY_10000000;
  base.lon = (float)ubxMessage.navPvt.lon * DEG_TO_RAD_BY_10000000;
  base.hMSL = (float)ubxMessage.navPvt.hMSL * 1e-3;
  // updade Matriz ECEF -> NED
  base.RotEcef2ned(0, 0) = -sinf(base.lat) * cosf(base.lon);
  base.RotEcef2ned(1, 0) = -sinf(base.lon);
  base.RotEcef2ned(2, 0) = -cosf(base.lat) * cosf(base.lon);
  base.RotEcef2ned(0, 1) = -sinf(base.lat) * sinf(base.lon);
  base.RotEcef2ned(1, 1) = cosf(base.lon);
  base.RotEcef2ned(2, 1) = -cosf(base.lat) * sinf(base.lon);
  base.RotEcef2ned(0, 2) = cosf(base.lat);
  base.RotEcef2ned(1, 2) = 0.0f;
  base.RotEcef2ned(2, 2) = -sinf(base.lat);
  while (processGPS() != MT_NAV_POSECEF) {
    vTaskDelay(1000);
  }
  updateNedPos();
#ifdef DEBUG_EKF
  printf("lon: %f\tlat: %f\thMSL: %f\t\n", base.lon, base.lat, base.hMSL);
  printf("%s\n", "Matrix de transformacao ECEF to NED");
  printEigen(base.RotEcef2ned);
  printf("%s\n", "Coordenadas ECEF");
  printf("%ld\t,%ld\t,%ld\t\n", base.rECEF[0] / 100, base.rECEF[1] / 100,
         base.rECEF[2] / 100);
  printf("%s\n", "Coordenadas NED em m");
  printEigen(rNEDgps);
#endif
  blindLED(5);
  turnOnLed();
}

/**
 * @brief Atualiza a posição ned de observação dada as informações contidas na
 * estrutura #base.
 */
void EKF::updateNedPos() {
  // NOTE: a mensagem ubxMessage.navPosecef fornece as medidas em cm.
  rNEDgps = base.RotEcef2ned.col(0) *
                (0.01 * (float)(ubxMessage.navPosecef.ecefX - base.rECEF[0])) +
            base.RotEcef2ned.col(1) *
                (0.01 * (float)(ubxMessage.navPosecef.ecefY - base.rECEF[1])) +
            base.RotEcef2ned.col(2) *
                (0.01 * (float)(ubxMessage.navPosecef.ecefZ - base.rECEF[2]));
  // atualiza a matriz H correspondente á posição.
  updateHp();
}
/**
 * @brief  Atualiza o vetor de velocidade no sistema NED #drNEDgps.
 */
void EKF::updateNedVel() {
  // NOTE: (float)ubxMessage.navPvt.velX está em mm/s.
  drNEDgps(0) = 0.001f * (float)ubxMessage.navPvt.velN;
  drNEDgps(1) = 0.001f * (float)ubxMessage.navPvt.velE;
  drNEDgps(2) = 0.001f * (float)ubxMessage.navPvt.velD;
  // atualiza a matriz H correspondente á velocidade
  updateHv();
}

/**
 * Metodos de calibracao
 * @param _porta_led [description]
 */
void EKF::calibrationMethods() {
  blindLED(3);
  imu.calibraGyro(200);
  blindLED(5);
  while (imu.calibracaoMagnetometro(22.8908f) < 0) {
    blindLED(2);
  }
  timeClass.computeElapsedTime();
  timeClass.computeElapsedTime();
  blindLED(3);
}
#endif

} // namespace ekf